File: scripts/algorithms/J/Jump Game VI/Jump Game VI.java
// Runtime: 20 ms (Top 93.03%) | Memory: 78 MB (Top 81.86%)
class Solution {
    public int maxResult(int[] nums, int k) {
        int n = nums.length, a = 0, b = 0;
        int[] deq = new int[n];
        deq[0] = n - 1;
        for (int i = n - 2; i >= 0; i--) {
            if (deq[a] - i > k) a++;
            nums[i] += nums[deq[a]];
            while (b >= a && nums[deq[b]] <= nums[i]) b--;
            deq[++b] = i;
        }
        return nums[0];
    }
}

File: scripts/algorithms/J/Jump Game VI/Jump Game VI.py
// Runtime: 791 ms (Top 67.87%) | Memory: 30.60 MB (Top 42.3%)

class Solution:
    def maxResult(self, nums, k):
        deq, n = deque([0]), len(nums)

        for i in range(1, n):
            while deq and deq[0] < i - k: deq.popleft()
            nums[i] += nums[deq[0]]   
            while deq and nums[i] >= nums[deq[-1]]: deq.pop()
            deq.append(i)
            
        return nums[-1]
